cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,302 @@
+        thrust::make_counting_iterator<size_type>(0), conv);
+
+    thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           unique_it,
+                           unique_it + input_size,
+                           dense_rank_sorted.data().get());
+  }
+  return dense_rank_sorted;
+}
+
+/**
+ * @brief Breaks the ties among equal value groups using binary operator and
+ * transform this tied value to final rank.
+ *
+ * @param dense_rank dense rank of sorted input column (acts as key for value
+ * groups).
+ * @param tie_iter  iterator of rank to break ties among equal value groups.
+ * @param sorted_order_view sorted order indices of input column
+ * @param rank_data output rank data
+ * @param tie_breaker tie breaking operator. For example, maximum & minimum.
+ * @param transformer transform after tie breaking (useful for average).
+ * @param stream stream to run the computations on
+ */
+template <typename TieType, typename outputType, typename TieBreaker, typename Transformer,
+          typename TieIterator>
+void tie_break_ranks_transform(rmm::device_vector<size_type> const &dense_rank_sorted,
+                               TieIterator tie_iter,
+                               column_view const &sorted_order_view,
+                               outputType *rank_data,
+                               TieBreaker tie_breaker,
+                               Transformer transformer,
+                               cudaStream_t stream) {
+  auto const input_size = sorted_order_view.size();
+  rmm::device_vector<TieType> tie_sorted(input_size, 0);
+  // algorithm: reduce_by_key(dense_rank, 1, n, reduction_tie_breaker)
+  // reduction_tie_breaker = min, max, min_count
+  thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
+                        dense_rank_sorted.begin(),
+                        dense_rank_sorted.end(),
+                        tie_iter,
+                        thrust::make_discard_iterator(),
+                        tie_sorted.begin(),
+                        thrust::equal_to<size_type>{},
+                        tie_breaker);
+  auto sorted_tied_rank = thrust::make_transform_iterator(
+      dense_rank_sorted.begin(),
+      [tied_rank = tie_sorted.begin(), transformer] __device__(auto dense_pos)
+          -> double { return transformer(tied_rank[dense_pos - 1]); });
+  thrust::scatter(rmm::exec_policy(stream)->on(stream),
+                  sorted_tied_rank,
+                  sorted_tied_rank + input_size,
+                  sorted_order_view.begin<size_type>(),
+                  rank_data);
+}
+
+template <typename outputType, typename TieBreaker>
+void tie_break_ranks(rmm::device_vector<size_type> const &dense_rank,
+                     column_view const &sorted_order_view,
+                     outputType *rank_data,
+                     TieBreaker tie_breaker,
+                     cudaStream_t stream) {
+  tie_break_ranks_transform<size_type, outputType>(
```

Comment:
Can this just be called directly from `rank_min/max` instead of needing a separate function?